saleor/warehouse/models.py

```diff
@@ -0,0 +1,62 @@
+import uuid
+from typing import Set
+
+from django.db import models
+from django.utils.translation import pgettext_lazy
+
+from ..account.models import Address
+from ..shipping.models import ShippingZone
+
+
+class WarehouseQueryset(models.QuerySet):
+    def prefetch_data(self):
+        return self.select_related("address").prefetch_related("shipping_zones")
+
+
+class Warehouse(models.Model):
+    id = models.UUIDField(default=uuid.uuid4, primary_key=True)
+    name = models.CharField(
+        pgettext_lazy("Warehouse field description", "Warehouse name"), max_length=255
+    )
+    company_name = models.CharField(
+        pgettext_lazy("Warehouse field description", "Legal company name"),
+        blank=True,
+        max_length=255,
+    )
+
+    shipping_zones = models.ManyToManyField(ShippingZone, blank=True)
+    address = models.ForeignKey(Address, on_delete=models.PROTECT)
+
+    email = models.EmailField(
+        pgettext_lazy("Warehouse field description", "Email address"),
+        blank=True,
+        default="",
+    )
+
+    objects = WarehouseQueryset.as_manager()
+
+    class Meta:
+        ordering = ("-name",)
+        permissions = (
+            (
+                "manage_warehouses",
+                pgettext_lazy("Permission description", "Manage warehouses."),
+            ),
+        )
+
+    def __str__(self):
+        return self.name
+
+    @property
+    def countries(self) -> Set[str]:
+        countries_zone = ",".join(
+            ShippingZone.objects.prefetch_related("warehouse_set")
+            .filter(warehouse=self)
+            .values_list("countries", flat=True)
+        )
+        return set(countries_zone.split(","))
```

Comment:
Join and split is some kind of force `str`? Or is there different purpose?

Reply:
Okay, I found that getting `shipping_zones.values_list("countries", flat=True)` returns a string with country code separated by commas and `shipping_zone.countries` returns a list of object :) I will add simplified version (that uses python objects) with next bach of commits :)